lib/prop/leaky_bucket_strategy.rb

```diff
@@ -0,0 +1,51 @@
+require 'prop/limiter'
+require 'prop/key'
+
+module Prop
+  class LeakyBucketStrategy
+    class << self
+      def update_bucket(cache_key, interval, leak_rate)
+        bucket = Prop::Limiter.reader.call(cache_key) || { :bucket => 0, :last_updated => 0 }
+        now = Time.now.to_i
+        leak_amount = (now - bucket[:last_updated]) / interval * leak_rate
+
+        bucket[:bucket] = [bucket[:bucket] - leak_amount, 0].max
+        bucket[:last_updated] = now
+
+        Prop::Limiter.writer.call(cache_key, bucket)
+      end
+
+      def counter(cache_key, options)
+        update_bucket(cache_key, options[:interval], options[:threshold])
```

Comment:
If `update_bucket` returns the new value of `bucket`, then you can assign `bucket = update_bucket(...)` here and not have to read back from the cache (which is another network operation)